packages/shared/scope-hoisting/src/concat.js

```diff
@@ -270,46 +280,66 @@ function wrapModule(asset: Asset, statements) {
           )) {
             decls.push(t.variableDeclarator(prop));
           }
-          if (init) {
-            body.push(
-              t.expressionStatement(t.assignmentExpression('=', id, init)),
-            );
-          }
         } else {
-          invariant(isIdentifier(id));
           decls.push(t.variableDeclarator(id));
-          let {init} = decl;
-          if (init) {
-            body.push(
-              t.expressionStatement(
-                t.assignmentExpression('=', t.identifier(id.name), init),
-              ),
+          invariant(t.isIdentifier(id));
+        }
+        if (
+          isForInStatement(parent, {left: node}) ||
+          isForOfStatement(parent, {left: node})
+        ) {
+          invariant(!init);
+          replace.push(id);
+        } else if (init) {
+          if (isForStatement(parent, {init: node})) {
+            replace.push(t.assignmentExpression('=', id, init));
+          } else {
+            replace.push(
+              t.expressionStatement(t.assignmentExpression('=', id, init)),
             );
           }
         }
       }
-    } else if (isFunctionDeclaration(node)) {
-      // Function declarations can be hoisted out of the module initialization function
-      fns.push(node);
-    } else if (isClassDeclaration(node)) {
-      let {id} = node;
-      invariant(isIdentifier(id));
-      // Class declarations are not hoisted. We declare a variable outside the
-      // function and convert to a class expression assignment.
-      decls.push(t.variableDeclarator(t.identifier(id.name)));
-      body.push(
-        t.expressionStatement(
-          t.assignmentExpression(
-            '=',
-            t.identifier(id.name),
-            t.toExpression(node),
-          ),
-        ),
-      );
+    }
+
+    if (replace.length > 1) {
+      path.replaceWithMultiple(replace).forEach(p => p.skip());
+    } else if (replace.length == 1) {
+      path.replaceWith(replace[0]);
+      path.skip();
     } else {
-      body.push(node);
+      path.remove();
     }
-  }
+  },
+  FunctionDeclaration(path, {fns}) {
+    fns.push(path.node);
+    path.remove();
+  },
+  FunctionExpression(path) {
+    path.skip();
+  },
+  ClassDeclaration(path, {decls}) {
+    let {node} = path;
+    let {id} = node;
+    invariant(isIdentifier(id));
+
+    // Class declarations are not hoisted. We declare a variable outside the
+    // function and convert to a class expression assignment.
+    decls.push(t.variableDeclarator(id));
+    path.replaceWith(
+      t.expressionStatement(
+        t.assignmentExpression('=', id, t.toExpression(node)),
+      ),
+    );
+    path.skip();
+  },
+};
+
+function wrapModule(asset: Asset, statements) {
+  let decls = [];
+  let fns = [];
+  let program = t.program(statements);
+  traverse(t.file(program), WRAP_MODULE_VISITOR, null, {decls, fns});
```

Comment:
What's the perf impact of this?

Reply:
I'll test it, but this is a `noScope` traverse that skips function and class bodies, so not too bad I think.

Reply:
Setup (similar to esbuild): 10 200k assets that each contain ~1000x `if(...){var foo1 = ...; let local = ...;}` and ~1000x `function bar1(){}`)

| | v2 | hoist-var |
|---|---|---|
| single `wrapModule` call | 2ms | 60ms
| all `wrapModule` "total time" | 18ms | 716ms
| build time | 14,7s | 15,8s

The question is now: how much of that is actually the overhead of the traverse and how much the fact the with v2, only half as many lines are processed.

Reply:
On the other hand, I don't see what we should do instead, the alternative would be to do a one-off reimplementation of babel-traverse. `babylon-traverse` doesn't really support removing nodes (because you don't have access to the parent and the position in the parent).

Reply:
That's a pretty significant difference, and I believe it's why I did a flat loop in the first place. But seems like it was insufficient, so it might be a cost we have to accept. We'll just have to find an equal or greater optimization elsewhere. :wink:

Did you profile to see what was expensive in there?

Reply:
> We'll just have to find an equal or greater optimization elsewhere.

Updating the npm ecosystem so wrapping isn't needed anymore...

> Did you profile to see what was expensive in there?

The visitor callbacks are red:
![Screenshot from 2020-03-21 22-49-27](https://user-images.githubusercontent.com/4586894/77237209-78a50a00-6bc6-11ea-826d-e56b0ed0ba93.png)

Reply:
It does seem to have a deeper visit depth than I expected. Are we missing a `path.skip()` somewhere?

Reply:
There isn't really anything to skip here:
```js
if (BigIntSupported) {
  var bigIntValue = () => true;
  let inner_let = false;
  if (inner_let) {
    throw new Error();
  }
}
// duplicated...

function testIsHoisted() {
  return BigIntSupported && bigIntValue();
}
// duplicated....
```

Reply:
How about all ExpressionStatements? We only need to process statements right?

Reply:
I've excluded all Expressions.

|  | small benchmark | bigger benchmark | [atlassian/parcel-stress-test](https://github.com/atlassian/parcel-stress-test)
|------|----|---|-----|
v2               | 5.7s | 18.2s | real: 1:30 - user: 3:05
custom traverse   | 5.0s | 20.1s | 1:22 - 3:00
hoist-var         | 5.7s | 20.8s | 1:24 - 3:00

"custom traverse" is a 80-line reimplmentaiton of babel-traverse that only does what we need here (`Path#replaceWith`, `Path#remove`, `Path#skip`).

Reply:
So in summary: this is an artificial benchmark and the `@babel/traverse` overhead is negligible in the real world.

Reply:
With the custom traverse, each of these traversals takes 28ms instead of 81ms, is that a significant difference?